### PR TITLE
shutdown: move busy mounts to not block parent mounts

### PR DESCRIPTION
**Pull request by valentindavid**
There is a case that confuses systemd-shutdown: a filesystem has been moved to a mount point which is part of another filesystem from an image from that former filesystem. systemd-shutdown cannot unmount any of those two filesystems. It needs first to move the filesystem containing the image of the other out of the tree of that image.
    
Here we move leaf mount points when they are busy so that they do not block parent mounts. We can only move leafs at each iteration since moving mount points also move sub mount points which would invalidate we read from

### Context

In Ubuntu Core we have have squashfs image that is mounted as rootfs by the boot initrd. This image is from a disk which is mounted within a mount point which, after switching root, is of course in the rootfs. So we cannot unmount the disk containing that the rootfs image. Though we can remount read-only.

Ubuntu has a package called finalrd which as a service, on `shutdown.target`, populates `/run/initramfs`. Effectively, `/run/initramfs/shutdown` is `systemd-shutdown`. So `systemd-shutdown` runs twice. Once in main rootfs. And then one in initrd. Then it is possible to move the mount containing the rootfs image out of `/oldroot`.

If we made `systemd-shutdown` smarter we would be able to get a safe shutdown without having to "wrap" `systemd-shutdown`.

### CODE REVIEW

**Comment by poettering** [comment]
hmm, can you give a simple example how an offending setup can look like? trying to wrap my head around this.

or let me try:

one file system mounted to `/foo/` and another one to `/foo/bar/`, and the `/foo/` file system is backed via loopback from a file `/foo/bar/quux`? Thus, when we try to unmount `/foo/` we can't because `/foo/bar/` is mounted inside it? And if we try to unmount `/foo/bar/` then we can't because the file `/foo/bar/quux` is used by the loopback device?

Did I get this right?

Your approach to fix this is to try to unmount them all, but if this doesn't work, then move them to a quarantine zone instead, where they can eventually be unmounted?

File: src/shutdown/umount.c [hunk 1]
```diff
@@ -45,6 +45,8 @@
 #include "umount.h"
 #include "util.h"
 #include "virt.h"
+#include "unit-name.h"
+#include "mkdir.h"
```
Comment (poettering) [inline src/shutdown/umount.c]: please keep this sorted

File: src/shutdown/umount.c [hunk 2]
```diff
@@ -172,12 +168,60 @@ int mount_points_list_get(const char *mountinfo, MountPoint **head) {
 
                 TAKE_PTR(remount_options);
 
+                unmountable = !nonunmountable_path(m->path);
+
+                // Order matters. We will remount from deep to root.
                 LIST_PREPEND(mount_point, *head, TAKE_PTR(m));
+        } while (false);
```
Comment (poettering) [inline src/shutdown/umount.c]: hmm, we'll leave the loop after the first iteration?

File: src/shutdown/umount.c [hunk 3]
```diff
@@ -172,12 +168,60 @@ int mount_points_list_get(const char *mountinfo, MountPoint **head) {
 
                 TAKE_PTR(remount_options);
 
+                unmountable = !nonunmountable_path(m->path);
+
+                // Order matters. We will remount from deep to root.
```
Comment (poettering) [inline src/shutdown/umount.c]: as per coding style: only use `/* */` style comments for explanatory stuff, and `//` for FIXME style stuff

File: src/shutdown/umount.c [hunk 4]
```diff
@@ -716,6 +753,32 @@ static int mount_points_list_umount(MountPoint **head, bool *changed, bool last_
                 if (nonunmountable_path(m->path))
                         continue;
 
+                if (keep_moving && m->should_move) {
+                        do {
+                                int r;
+                                _cleanup_free_ char *newpath = NULL;
+                                _cleanup_free_ char *escaped = NULL;
+
+                                r = unit_name_path_escape(m->path, &escaped);
```
Comment (poettering) [inline src/shutdown/umount.c]: this cannot work. PATH_MAX is 4K, but NAME_MAX is 256.

This low-level stuff (unike the high-level units) really should work for all paths, hence you need to find a different way. (for example, just hash the stuff to a sufficiently large hash value. we have siphas and sha256 anyway9

File: src/shutdown/umount.c [hunk 5]
```diff
@@ -716,6 +753,32 @@ static int mount_points_list_umount(MountPoint **head, bool *changed, bool last_
...
+                                if (r != 0) {
+                                        keep_moving = false;
+                                        break ;
+                                }
+                                asprintf(&newpath, "/run/shutdown/mounts/%s", escaped);
```
Comment (poettering) [inline src/shutdown/umount.c]: oom check missing. also, path_join()

**Comment by valentindavid** [comment]
> one file system mounted to `/foo/` and another one to `/foo/bar/`, and the `/foo/` file system is backed via loopback from a file `/foo/bar/quux`? Thus, when we try to unmount `/foo/` we can't because `/foo/bar/` is mounted inside it? And if we try to unmount `/foo/bar/` then we can't because the file `/foo/bar/quux` is used by the loopback device?
> 
> Did I get this right?

Correct.

> Your approach to fix this is to try to unmount them all, but if this doesn't work, then move them to a quarantine zone instead, where they can eventually be unmounted?

Not exactly. In my change,  we mark unmountable mountpoints that are under other unmountable mount points. The marked mountpoints will be all moved before we unmount them. So that if they are not unmounted, they are at least moved and not blocking parent mountpoints.

By unmountable mountpoints, I mean all mountpoints that is listed by `mount_points_list_get_iter` (so not `/run` or `/dev` for example) for which `nonunmountable_path` is false (so not `/` for example).

File: src/shutdown/umount.c (hunk 4 above)
Comment (valentindavid) [reply src/shutdown/umount.c]: OK. Do you have suggestion? I can use a random temporary name. But we lose information on what the original mountpoint was if there are errors later.

File: src/shutdown/umount.c (hunk 2 above)
Comment (valentindavid) [reply src/shutdown/umount.c]: I can change the `continue`s to a `goto` instead. It will be less confusing.

**Comment by valentindavid** [comment]
> so, can't we make this a systematic strategy: whenever we get EBUSY on umount() let's try to move the mount to /run/shutdown/umount/$SOMETHING or so (unless it already is there), and then continue with the umount loop as if we had unmounted it? In the next iteration of the outer loop we'll then see the mount on the new place and unmount it from there?
> 
> that should be a much simpler patch, no?

Not that simple. Moving a mount point that contains other mount points move those sub mount points. That means if we continue iterating on the list we collected, the `m->path` may be invalid. So we will get ENOENT (or EINVAL) on other umounts or remounts.

Do we care? Should we accept doing more invalid system calls but have simple code? Or should we calculate the proper order (that is from deepest mount points)?

File: src/shutdown/umount.c (hunk 4 above)
Comment (valentindavid) [reply src/shutdown/umount.c]: I will use sha256

**Comment by poettering** [comment]
hmm, MS_MOVE is always recursive, I see. But we could manually check if it is a leaf mount? And only then move it away then?

**Comment by valentindavid** [comment]
> hmm, MS_MOVE is always recursive, I see. But we could manually check if it is a leaf mount? And only then move it away then?

So when we list the mounts, we could call `mnt_table_next_child_fs` and if does not give us any child, we mark it as leaf.

When unmounting, if we get EBUSY and it is marked as leaf, we move it. We also mark `*changed = true` to iterate more.

I think that should work. I will test that.

**Comment by valentindavid** [comment]
I have successfully tested this commit:  https://github.com/valentindavid/systemd/commit/fc7371c2cef97f70cad4320127d1f6810faf26e4

Should I force push in this branch? Or should I create a new PR?

**Comment by bluca** [comment]
> I have successfully tested this commit: [valentindavid@fc7371c](https://github.com/valentindavid/systemd/commit/fc7371c2cef97f70cad4320127d1f6810faf26e4)
> 
> Should I force push in this branch? Or should I create a new PR?

just push here

**Comment by valentindavid** [comment]
> just push here

Done. Also updated the title and the description of the PR.

File: src/shutdown/umount.c [hunk 6]
```diff
@@ -155,6 +160,22 @@ int mount_points_list_get(const char *mountinfo, MountPoint **head) {
...
+
+                struct libmnt_fs *child;
+                r = mnt_table_next_child_fs(table, iter_children, fs, &child);
+                if (r < 0) {
+                        return log_error_errno(r, "Failed to get children mounts for %s from %s: %m", path, mountinfo ?: "/proc/self/mountinfo");
+                } else if (r == 1) {
+                        leaf = true;
+                } else {
+                        leaf = false;
+                }
```
Comment (keszybz) [inline src/shutdown/umount.c]:
```C
if (r < 0)
    return ...;
bool leaf = r == 1;
```

No {} around single-line blocks.

File: src/shutdown/umount.c [hunk 7]
```diff
@@ -155,6 +160,22 @@ int mount_points_list_get(const char *mountinfo, MountPoint **head) {
...
+                /* We care only whether it exists, it is unused */
+                (void)child;
```
Comment (keszybz) [inline src/shutdown/umount.c]: Hmm, what is this code doing? If this is about silencing of a warning about an unused variable, put `_unused_` on the declaration instead.

File: src/shutdown/umount.c [hunk 8]
```diff
@@ -716,10 +739,53 @@ static int mount_points_list_umount(MountPoint **head, bool *changed, bool last_
...
                 else
                         *changed = true;
+
+                /* If a mount is busy, we move it to not keep parent mount points busy.
+                 * If a mount point is not a leaf, moving it would invalidate our mount table.
+                 * More moving will occur in next iteration with a fresh mount table.
+                 *
+                 * umount_with_timeout will return EPROTO even on EBUSY due to forking.
+                 */
+                if (r == -EPROTO && m->leaf) {
```
Comment (keszybz) [inline src/shutdown/umount.c]: Maybe invert this to reduce indentation of the remaining code?
```c
if (r != -EPROTO || !m->leaf)
       continue;
```
?

File: src/shutdown/umount.c [hunk 9]
```diff
@@ -716,10 +739,53 @@ static int mount_points_list_umount(MountPoint **head, bool *changed, bool last_
...
+
+                                sha256_direct(m->path, strlen(m->path), hash);
+
+                                hexhash = hexmem(hash, SHA256_DIGEST_SIZE);
+                                if (!hexhash)
+                                        return log_oom();
+
+                                asprintf(&newpath, "/run/shutdown/mounts/%s", hexhash);
+                                if (!newpath)
+                                        return log_oom();
```
Comment (keszybz) [inline src/shutdown/umount.c]: That is both overcomplicated and not enough. It is possible to stack multiple mounts will be on the same path, which would result in the same hash and then we'd be stacking the mounts, but in reverse order. Just maintain a counter:
```C
char newpath[STRLEN("/run/shutdown/mounts/" + DECIMAL_STRING_MAX];
xsprintf(newpath, "/run/shutdown/mounts/%u", n_moved);

...

n_moved ++;
```

File: src/shutdown/umount.c [hunk 10]
```diff
@@ -716,10 +739,53 @@ static int mount_points_list_umount(MountPoint **head, bool *changed, bool last_
...
+
+                                (void) mkdir_p(newpath, 0700);
+
+                                r = mount(m->path, newpath, NULL, MS_MOVE, NULL);
+
```
Comment (keszybz) [inline src/shutdown/umount.c:807]: No empty line between action and error handling.

File: src/shutdown/umount.c [hunk 11]
```diff
@@ -716,10 +739,53 @@ static int mount_points_list_umount(MountPoint **head, bool *changed, bool last_
...
+                                if (!hexhash)
+                                        return log_oom();
+
+                                asprintf(&newpath, "/run/shutdown/mounts/%s", hexhash);
+                                if (!newpath)
+                                        return log_oom();
+
+                                (void) mkdir_p(newpath, 0700);
+
+                                r = mount(m->path, newpath, NULL, MS_MOVE, NULL);
```
Comment (keszybz) [inline src/shutdown/umount.c]: Please write this as `r = NERRNO(mount(…));`
and then just use `if (r < 0)` below.

File: src/shutdown/umount.h [hunk 12]
```diff
@@ -20,6 +20,7 @@ typedef struct MountPoint {
         unsigned long remount_flags;
         bool try_remount_ro:1;
         bool umount_lazily:1;
+        bool leaf:1;
```
Comment (keszybz) [inline src/shutdown/umount.h]: Oh, that's something to fix independently. Please make a separate commit to drop the bitfield markers. They make the code text more complicated but don't actually save any space:
```
struct MountPoint {
        char *                     path;                 /*     0     8 */
        char *                     remount_options;      /*     8     8 */
        long unsigned int          remount_flags;        /*    16     8 */
        _Bool                      try_remount_ro:1;     /*    24: 0  1 */
        _Bool                      umount_lazily:1;      /*    24: 1  1 */

        /* XXX 6 bits hole, try to pack */
        /* XXX 7 bytes hole, try to pack */

        dev_t                      devnum;               /*    32     8 */
```

**Review by keszybz — CHANGES_REQUESTED** [review]

File: src/shutdown/umount.c [hunk 13]
```diff
@@ -155,6 +160,22 @@ int mount_points_list_get(const char *mountinfo, MountPoint **head) {
...
 
+                iter_children = mnt_new_iter(MNT_ITER_FORWARD);
+                if (!iter_children)
+                        return log_oom();
+
+                struct libmnt_fs *child;
+                r = mnt_table_next_child_fs(table, iter_children, fs, &child);
+                if (r < 0) {
+                        return log_error_errno(r, "Failed to get children mounts for %s from %s: %m", path, mountinfo ?: "/proc/self/mountinfo");
+                } else if (r == 1) {
```
Comment (poettering) [inline src/shutdown/umount.c]: no {} around single-line if blocks, as per coding style.

also: given the first `if` branch here exits the function you can drop the `else` without change of effect, it's redundant

File: src/shutdown/umount.c [hunk 14]
```diff
@@ -716,10 +739,53 @@ static int mount_points_list_umount(MountPoint **head, bool *changed, bool last_
...
+                                uint8_t hash[SHA256_DIGEST_SIZE];
+
+                                sha256_direct(m->path, strlen(m->path), hash);
+
+                                hexhash = hexmem(hash, SHA256_DIGEST_SIZE);
+                                if (!hexhash)
+                                        return log_oom();
+
+                                asprintf(&newpath, "/run/shutdown/mounts/%s", hexhash);
+                                if (!newpath)
```
Comment (poettering) [inline src/shutdown/umount.c]: this is likely going to trigger a static checker sooner or later, because you ignore the asprintf() return value here. hence do:

```c
if (asprintf(…) < 0)
        return log_oom();
```

File: src/shutdown/umount.c (hunk 9 above)
Comment (poettering) [reply src/shutdown/umount.c]: counter would have to be static...

honestly I'd probably just use a random name. i.e. just use random_u64()

Comment (valentindavid) [reply src/shutdown/umount.c]: It would be fine if they got mounted in the same place. The important thing is that they do no get mounted in a subdirectory.

I can use a random number instead to make it simpler.

But using a hash was mainly to have reproducible error messages if there are failures to umount further. And it would allow, from a list of mount point, hash them and figure out which one is problematic. With random numbers or a counter, it will be very difficult to know what mount is problematic.

File: src/shutdown/umount.c (hunk 14 above)
Comment (valentindavid) [reply src/shutdown/umount.c]: I have replaced it with `xsprintf` like proposed by @keszybz

File: src/shutdown/umount.c (hunk 13 above)
Comment (valentindavid) [reply src/shutdown/umount.c]: Done

File: src/shutdown/umount.h (hunk 12 above)
Comment (valentindavid) [reply src/shutdown/umount.h]: Done.

File: src/shutdown/umount.c (hunk 11 above)
Comment (valentindavid) [reply src/shutdown/umount.c]: Done. I have also found that few lines before, I was using `errno` instead of `r` when reporting error. I have fixed that.

File: src/shutdown/umount.c (hunk 10 above)
Comment (valentindavid) [reply src/shutdown/umount.c:807]: Done.

File: src/shutdown/umount.c (hunk 8 above)
Comment (valentindavid) [reply src/shutdown/umount.c]: Sure, that looks better. Done.

File: src/shutdown/umount.c (hunk 7 above)
Comment (valentindavid) [reply src/shutdown/umount.c]: Correct. Done.

File: src/shutdown/umount.c (hunk 6 above)
Comment (valentindavid) [reply src/shutdown/umount.c]: Done.

File: src/shutdown/umount.c [hunk 15]
```diff
@@ -155,6 +158,18 @@ int mount_points_list_get(const char *mountinfo, MountPoint **head) {
...
+                iter_children = mnt_new_iter(MNT_ITER_FORWARD);
+                if (!iter_children)
+                        return log_oom();
+
+                /* We care only whether it exists, it is unused */
+                _unused_ struct libmnt_fs *child;
+                r = mnt_table_next_child_fs(table, iter_children, fs, &child);
+                if (r < 0) {
+                        return log_error_errno(r, "Failed to get children mounts for %s from %s: %m", path, mountinfo ?: "/proc/self/mountinfo");
+                }
```
Comment (keszybz) [inline src/shutdown/umount.c]: No {} around single lines please.

File: src/shutdown/umount.c [hunk 16]
```diff
@@ -716,10 +733,44 @@ static int mount_points_list_umount(MountPoint **head, bool *changed, bool last_
...
+
+                r = path_extract_directory(m->path, &dirname);
+                if (r < 0) {
+                        n_failed++;
+                        log_full_errno(last_try ? LOG_ERR : LOG_INFO, r, "Cannot find directory for %s: %m", m->path);
+                        continue;
+                }
+
+                if (!path_equal_or_files_same(dirname, "/run/shutdown/mounts", 0)) {
+                        char newpath[STRLEN("/run/shutdown/mounts/") + DECIMAL_STR_MAX(uint64_t) + 1];
```
Comment (keszybz) [inline src/shutdown/umount.c]: Why `+1`? If this is about the terminator, please drop it. `DECIMAL_STR_MAX` sadly already includes `+1` internally.

File: src/shutdown/umount.c [hunk 17]
```diff
@@ -716,10 +733,44 @@ static int mount_points_list_umount(MountPoint **head, bool *changed, bool last_
...
+
+                        xsprintf(newpath, "/run/shutdown/mounts/%" PRIu64, random_u64());
```
Comment (keszybz) [inline src/shutdown/umount.c]: Maybe `"%08" PRIx64` instead? Otherwise we'll get different lengths, which will look moderately ugly.

**Review by keszybz — COMMENTED** [review]
I think the random paths are OK. If the user should know what path is causing a problem, then we should improve logging. For example, we could log the original mount point, or the device name, etc. The temporary mount point should not be relevant.

File: src/shutdown/umount.c [hunk 18]
```diff
@@ -682,6 +698,7 @@ static int umount_with_timeout(MountPoint *m, bool last_try) {
  * this function is invalidated, and should not be reused. */
 static int mount_points_list_umount(MountPoint **head, bool *changed, bool last_try) {
         int n_failed = 0;
+        int r;
```
Comment (poettering) [inline src/shutdown/umount.c]: maybe merge with previous line?

File: src/shutdown/umount.c [hunk 19]
```diff
@@ -716,10 +733,44 @@ static int mount_points_list_umount(MountPoint **head, bool *changed, bool last_
...
+                        log_full_errno(last_try ? LOG_ERR : LOG_INFO, r, "Cannot find directory for %s: %m", m->path);
+                        continue;
+                }
+
+                if (!path_equal_or_files_same(dirname, "/run/shutdown/mounts", 0)) {
+                        char newpath[STRLEN("/run/shutdown/mounts/") + DECIMAL_STR_MAX(uint64_t) + 1];
+
+                        xsprintf(newpath, "/run/shutdown/mounts/%" PRIu64, random_u64());
+
+                        (void) mkdir_p(newpath, 0700);
```
Comment (poettering) [inline src/shutdown/umount.c]: hmm, this assumes the mounts we move away here are dirs. they could be non-dir inodes too. Those you cannot mount on top of a dir inode. (you can mount dir inodes on dir inodes only, and non-dir on any non-dir)

hence, stat() the source first, and then either do am mkdir or a touch() to create either a dir or a regular file.

Comment (poettering) [reply src/shutdown/umount.c]: (and yeah, i think it is worth getting this right, thinkg a root fs mounted off a loopback file whose backing fs is mounted somewhere at the leaf of the tree)

**Comment by poettering** [comment]
looks really good, but one corner case needs addressing I think

File: src/shutdown/umount.c (hunk 19 above)
Comment (valentindavid) [reply src/shutdown/umount.c]: Done. I also added error handling to calls to mkdir_p and touch_file in order to make the logs more obvious.

File: src/shutdown/umount.c [hunk 20]
```diff
@@ -155,6 +159,17 @@ int mount_points_list_get(const char *mountinfo, MountPoint **head) {
...
+                iter_children = mnt_new_iter(MNT_ITER_FORWARD);
+                if (!iter_children)
+                        return log_oom();
+
+                /* We care only whether it exists, it is unused */
+                _unused_ struct libmnt_fs *child;
+                r = mnt_table_next_child_fs(table, iter_children, fs, &child);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to get children mounts for %s from %s: %m", path, mountinfo ?: "/proc/self/mountinfo");
+                bool leaf = r == 1;
```
Comment (poettering) [inline src/shutdown/umount.c]: btw, looking at this again I think it would be fantastic to make this a helper func, maybe moved to src/shared/libmount-util.[ch], called `libmount_is_leaf()`

File: src/shutdown/umount.c [hunk 21]
```diff
@@ -716,10 +732,49 @@ static int mount_points_list_umount(MountPoint **head, bool *changed, bool last_
...
+                if (r < 0)
                         n_failed++;
                 else
                         *changed = true;
+
+                /* If a mount is busy, we move it to not keep parent mount points busy.
+                 * If a mount point is not a leaf, moving it would invalidate our mount table.
+                 * More moving will occur in next iteration with a fresh mount table.
+                 *
+                 * umount_with_timeout will return EPROTO even on EBUSY due to forking.
```
Comment (poettering) [inline src/shutdown/umount.c]: we should fix that btw, i.e. send the error to the parent process. → #25987 

(could you please rebase on that [or maybe just pull that commit into your PR], and then check for EBUSY explicitly here?)

Comment (poettering) [reply src/shutdown/umount.c]: please suffix function names in docs/comments with `()` to indicate these are functions.

File: src/shutdown/umount.c [hunk 22]
```diff
@@ -716,10 +732,49 @@ static int mount_points_list_umount(MountPoint **head, bool *changed, bool last_
...
+                if (r < 0) {
+                        n_failed++;
+                        log_full_errno(last_try ? LOG_ERR : LOG_INFO, r, "Cannot find directory for %s: %m", m->path);
+                        continue;
+                }
+
+                if (!path_equal_or_files_same(dirname, "/run/shutdown/mounts", 0)) {
+                        char newpath[STRLEN("/run/shutdown/mounts/") + DECIMAL_STR_MAX(uint64_t)];
+
+                        xsprintf(newpath, "/run/shutdown/mounts/%08" PRIu64, random_u64());
```
Comment (poettering) [inline src/shutdown/umount.c]: hmm, so you use DECIMAL_STR_MAX() above to size the buffer, but hten you hardcode a literal size "8" here after all. either use the `*` sprintf() thing, to also use that here. or just drop the `08`. or just use hex output, then the size is clear...

(i'd personally just go for hex)

File: src/shutdown/umount.c [hunk 23]
```diff
@@ -716,10 +732,49 @@ static int mount_points_list_umount(MountPoint **head, bool *changed, bool last_
...
+
+                        if (is_dir(m->path, true))
+                                (void) mkdir_p(newpath, 0700);
```
Comment (poettering) [inline src/shutdown/umount.c]: i'd make the access mode 000 btw, it's not supposed to be accessed after all, the overmounted fs should be the only thing

File: src/shutdown/umount.c [hunk 24]
```diff
@@ -716,10 +732,49 @@ static int mount_points_list_umount(MountPoint **head, bool *changed, bool last_
...
+                        else
+                                (void) touch_file(newpath, true, USEC_INFINITY, UID_INVALID, GID_INVALID, 0700);
```
Comment (poettering) [inline src/shutdown/umount.c]: same here

also, please prefix the second argument with `/* parents= */` as we do these days in our codebase

File: src/shutdown/umount.c [hunk 25]
```diff
@@ -744,9 +747,25 @@ static int mount_points_list_umount(MountPoint **head, bool *changed, bool last_
...
-                if (r != -EPROTO || !m->leaf)
+                if (r != -EPROTO)
                         continue;
 
+                if  (!m->leaf) {
+                        /* If not a leaf and busy, it might be because some submounts are busy and need to be
+                         * moved out. We can only move mounts whose parent mounts are private.
+                         */
+                        if (m->shared) {
+                                r = RET_NERRNO(mount(NULL, m->path, NULL, MS_PRIVATE, NULL));
```
Comment (poettering) [inline src/shutdown/umount.c]: hmm, i think it would be better to just remount *everything* MS_PRIVATE as an early step during shutdown. i.e. somewhere close to the mlockall() call in shutdown.c do an MS_PRIVATE|MS_REC call on "/", so that everything is made private. Trying to do this individually is just needlessly fragile I'd say.

At that time we really don't want to allow multiple namespaces or propagation anymore, and not be confused by its effects, and hence let's just turn if off altogether.

File: src/shutdown/umount.c [hunk 26]
```diff
@@ -761,10 +780,19 @@ static int mount_points_list_umount(MountPoint **head, bool *changed, bool last_
...
+                                        log_full_errno(last_try ? LOG_ERR : LOG_INFO, r, "Could not create directory %s: %m", newpath);
+                                        continue;
+                                }
+                        } else {
+                                r = touch_file(newpath, true, USEC_INFINITY, UID_INVALID, GID_INVALID, 0700);
+                                if (r < 0) {
+                                        log_full_errno(last_try ? LOG_ERR : LOG_INFO, r, "Could not create file %s: %m", newpath);
+                                        continue;
+                                }
+                        }
```
Comment (poettering) [inline src/shutdown/umount.c:804]: this part of the change ended up in the wrong commit i guess?

**Review by poettering — CHANGES_REQUESTED** [review]

File: src/shutdown/umount.c (hunk 25 above)
Comment (valentindavid) [reply src/shutdown/umount.c]: I suppose we should do that only when not in a container.

Comment (poettering) [reply src/shutdown/umount.c]: why? i think we can do that anywhere.

**Comment by poettering** [comment]
https://github.com/systemd/systemd/pull/25987 is merged now. please rebase, and then check for EBUSY explicitly

File: src/shutdown/umount.c (hunk 26 above)
Comment (valentindavid) [reply src/shutdown/umount.c:804]: It was.

File: src/shutdown/umount.c (hunk 25 above)
Comment (valentindavid) [reply src/shutdown/umount.c]: Done. 

This revealed a bug in testing, where I tested if a mount was already moved. I was using `path_equal_or_files_same`. But if `/run` was bind mounted somewhere else, that did not work. When re-running `systemd-shutdown` in the shutdown initramfs, mounts in `/oldroot/run/shutdown/mounts` were not being moved to `/run/shutdown/mounts` which was blocking unmount of `/oldroot`.

I have changed that to using `chase_symlinks` on `/run/shutdown/mounts` and comparing the mount point to that.

File: src/shutdown/umount.c (hunk 24 above)
Comment (valentindavid) [reply src/shutdown/umount.c]: Done.

File: src/shutdown/umount.c (hunk 23 above)
Comment (valentindavid) [reply src/shutdown/umount.c]: Done.

File: src/shutdown/umount.c (hunk 22 above)
Comment (valentindavid) [reply src/shutdown/umount.c]: I changed it to hex.

File: src/shutdown/umount.c (hunk 21 above)
Comment (valentindavid) [reply src/shutdown/umount.c]: It is rebased.

File: src/shutdown/umount.c (hunk 20 above)
Comment (valentindavid) [reply src/shutdown/umount.c]: Done.

**Comment by keszybz** [comment]
```
../src/shutdown/umount.c: In function ‘mount_points_list_umount’:
../src/shutdown/umount.c:784:32: error: implicit declaration of function ‘chase_symlinks’; did you mean ‘is_symlink’? [-Werror=implicit-function-declaration]
  784 |                         (void) chase_symlinks("/run/shutdown/mounts", NULL, 0, &resolved_mounts_path, NULL);
      |                                ^~~~~~~~~~~~~~
      |                                is_symlink
../src/shutdown/umount.c:784:32: error: nested extern declaration of ‘chase_symlinks’ [-Werror=nested-externs]
```

**Comment by keszybz** [comment]
Looks very nice.

An RFE, since this needs another round anyway: please put the justifications for the changes in the commits. Both 1/3 and 3/3 have no information except the title. It really really helps when there's some bug or unexpected interaction to be able to see what was the goal of the commit. And with changes so deep in the logic, it's likely that there will be some follow-up.

File: src/shared/libmount-util.c [hunk 27]
```diff
@@ -38,3 +38,23 @@ int libmount_parse(
...
+
+int libmount_is_leaf(
+                struct libmnt_table *table,
+                struct libmnt_fs *fs) {
+        int r;
+
+        _cleanup_(mnt_free_iterp) struct libmnt_iter *iter_children = NULL;
+        iter_children = mnt_new_iter(MNT_ITER_FORWARD);
+
+        if (!iter_children)
```
Comment (poettering) [inline src/shared/libmount-util.c:49]: please drop the empty line here. we generally have the rule to keep function calls and their immediate error handling together. hence maybe put an empty line before the line with the mnt_new_iter() call, but not immediately after, as the `if` here is then the error handling

File: src/shared/libmount-util.c [hunk 28]
```diff
@@ -38,3 +38,23 @@ int libmount_parse(
...
+                return log_oom();
+
+        /* We care only whether it exists, it is unused */
+        _unused_ struct libmnt_fs *child;
+        r = mnt_table_next_child_fs(table, iter_children, fs, &child);
```
Comment (poettering) [inline src/shared/libmount-util.c:54]: i wished mnt_table_next_child_fs() whould just allow us to pass NULL as last argument here, but i guess it does not. :-(

File: src/shutdown/umount.c [hunk 29]
```diff
@@ -744,10 +755,62 @@ static int mount_points_list_umount(MountPoint **head, bool *changed, bool last_
...
+                        n_failed++;
+                        log_full_errno(last_try ? LOG_ERR : LOG_INFO, r, "Cannot find directory for %s: %m", m->path);
+                        continue;
+                }
+
+                /* We need to canonicalize /run/shutdown/mounts. We cannot compare inodes, since /run
+                 * might be bind mounted somewhere we want to unmount. And we need to move all mounts in
+                 * /run/shutdown/mounts from there.
+                 */
+                if (resolved_mounts_path == NULL)
```
Comment (poettering) [inline src/shutdown/umount.c]: when it comes to pointers we usually do NULL checks relying on C's downgrade-to-bool checks, i.e.

```c
if (!resolved_mounts_path)
```

but doesn't matter much

File: src/shutdown/umount.c [hunk 30]
```diff
@@ -744,10 +755,62 @@ static int mount_points_list_umount(MountPoint **head, bool *changed, bool last_
...
+                        (void) chase_symlinks("/run/shutdown/mounts", NULL, 0, &resolved_mounts_path, NULL);
+                if ((resolved_mounts_path == NULL) || !path_equal(dirname, resolved_mounts_path)) {
```
Comment (poettering) [inline src/shutdown/umount.c]: path_equal() is actually fine with NULL params. hence you can shorten this to just:

```c
if (!path_equal(dirname, resolved_mounts_path))
``

File: src/shutdown/umount.c [hunk 31]
```diff
@@ -744,10 +755,62 @@ static int mount_points_list_umount(MountPoint **head, bool *changed, bool last_
...
+                        char newpath[STRLEN("/run/shutdown/mounts/") + 16 + 1];
+
+                        xsprintf(newpath, "/run/shutdown/mounts/%016" PRIx64, random_u64());
+
+                        if (is_dir(m->path, true)) {
```
Comment (poettering) [inline src/shutdown/umount.c]: is_dir() can return negative on error it's probably fine to ignore that, but if you do, please make this clear by saying `is_dir(…) != 0`

i.e. if you drop the `!= 0` the next one reading this might believe it returns a boolean, but it doesn't do that, it returns an "extended boolean", i.e. a boolean or negative error.

File: src/shutdown/shutdown.c [hunk 32]
```diff
@@ -400,6 +400,9 @@ int main(int argc, char *argv[]) {
         /* Lock us into memory */
         (void) mlockall(MCL_CURRENT|MCL_FUTURE);
 
+        if (mount(NULL, "/", NULL, MS_REC|MS_PRIVATE, NULL) < 0)
```
Comment (poettering) [inline src/shutdown/shutdown.c:405]: please add a brief comment here as to why. i.e just say "we want to be able to potentially use MS_MOVE later, and that only works on MS_PRIVATE mounts" or so.

File: src/shutdown/shutdown.c [hunk 33]
```diff
@@ -400,6 +400,9 @@ int main(int argc, char *argv[]) {
...
+                log_warning_errno(errno, "Failed to make mounts private: %m");
```
Comment (poettering) [inline src/shutdown/shutdown.c]: we try to suffix log messages about errors we ignore with `, ignoring`. i.e. maybe change this to:

```
log_warning_errno(errno, "Failed to make mounts private, ignoring: %m");
```

**Comment by poettering** [comment]
some nitpicky stuff left. looks excellent otherwise, i guess we can merge this soon.

**Comment by poettering** [comment]
> An RFE, since this needs another round anyway: please put the justifications for the changes in the commits. Both 1/3 and 3/3 have no information except the title. It really really helps when there's some bug or unexpected interaction to be able to see what was the goal of the commit. And with changes so deep in the logic, it's likely that there will be some follow-up.

I actually would prefer the key info as comments of the code, rather than in the comments in the commit msg. → https://systemd.io/CODING_STYLE/#commenting

File: src/shared/libmount-util.c (hunk 28 above)
Comment (valentindavid) [reply src/shared/libmount-util.c:54]: Looking at the source, it does de-reference it even if NULL.

Comment (poettering) [reply src/shared/libmount-util.c:54]: → https://github.com/util-linux/util-linux/issues/2039

Comment (poettering) [reply src/shared/libmount-util.c:54]: our wishes were heard by karel. But I guess we need to retain compat with old libmount versions for a longer time, so nothing to change on our side
